source/common/upstream/conn_pool_map.h

```diff
@@ -0,0 +1,59 @@
+#pragma once
+
+#include <map>
+#include <vector>
+
+#include "envoy/event/dispatcher.h"
+
+#include "absl/types/optional.h"
+
+namespace Envoy {
+namespace Upstream {
+/**
+ *  A class mapping keys to connection pools, with some recycling logic built in.
+ */
+template <typename KEY_TYPE, typename POOL_TYPE> class ConnPoolMap {
+public:
+  using PoolFactory = std::function<std::unique_ptr<POOL_TYPE>()>;
+  using DrainedCb = std::function<void()>;
+
+  ConnPoolMap(Event::Dispatcher& dispatcher);
+  ~ConnPoolMap();
+  /**
+   * Returns an existing pool for `key`, or creates a new one using `factory`. Note that it is
+   * possible for this to fail if a limit on the number of pools allowed is reached.
+   * @return The pool corresponding to `key`, or `absl::nullopt`
+   */
+  absl::optional<POOL_TYPE*> getPool(KEY_TYPE key, PoolFactory factory);
+
+  /**
+   * @return the number of pools
+   */
+  size_t size() const;
+
+  /**
+   * Destroys all mapped pools.
+   */
+  void clear();
+
+  /**
+   * Adds a drain callback to all mapped pools. Any future mapped pools with have the callback
+   * automatically added. Be careful with the callback. If it itself calls into `this`, modifying
+   * the state of `this`, there is a good chance it will cause corruption due to the callback firing
+   * immediately.
+   */
+  void addDrainedCallback(DrainedCb cb);
+
+  /**
+   * Instructs each connection pool to drain its connections
+   */
+  void drainConnections();
+
+private:
+  std::map<KEY_TYPE, std::unique_ptr<POOL_TYPE>> active_pools_;
```

Comment:
unordered_map?

Reply:
Hehe, I was thinking of that, but got lazy defining a hash key for the vector. I'll think about it a bit. Note: the original map was `std::map`. That said, now that we're looking to use this more heavily, we'll likely want a more efficient structure.

I'll try out unordered_map. If it's complicated I may do another PR for it.

Reply:
Sorry I haven't been watching the context around this PR. But, you may find it "just works" to use `absl::flat_hash_map` with a vector as a key, assuming the elements of the vector are themselves hashable. See https://abseil.io/docs/cpp/guides/hash

Reply:
@dnoe thanks for the suggestion. I'll give it a try.

Reply:
@dnoe It seems to work. Looks like it ends up using https://github.com/abseil/abseil-cpp/blob/d78310fe5a82f2e0e6e16509ef8079c8d7e4674e/absl/hash/internal/hash.h#L411. Thanks again.